Triggers query location use case for user email

sven1103 opened the pull request:
- [x] show available locations from the beginning #8 
- [x] link view action to controller
- [x] call use case from controller
- [x] implement use case output in the presenter

sven1103 committed:
Triggers query location use case for user email

KochTobi commented:
Address #8

sven1103 committed:
Implements default constructor for empty view model

sven1103 committed:
Removes unused imports

KochTobi committed:
prevent default constructor for ControlElements

sven1103 committed:
Fix the malfunctioning location query

sven1103 committed:
Merge branch 'feature/usecase-controlelements' of github.com:qbicsoftware/sample-tracking-update-portlet into feature/usecase-controlelements

sven1103 committed:
Loads the available locations in the selection list

sven1103 committed:
Removes unnecessary empty selection caption

sven1103 committed:
Adds location selection caption

KochTobi commented on src/main/groovy/life/qbic/portal/sampletracking/web/views/ControlElements.groovy:
```diff
@@ -20,25 +21,34 @@ class ControlElements extends VerticalLayout {
+        locationSelectMenu.setCaption("Available locations")
+        locationSelectMenu.emptySelectionAllowed = false
+        locationSelectMenu.setDataProvider(new ListDataProvider<>(viewModel.availableLocations))
+        locationSelectMenu.setItemCaptionGenerator({it -> it.name})
```
```suggestion
        locationSelectMenu.setItemCaptionGenerator({it -> it?.name})
```
Here the NullSafeOperator could makes sense.

sven1103 committed:
Removes unnecessary code

KochTobi commented on src/main/groovy/life/qbic/portal/sampletracking/DependencyManager.groovy:
```diff
@@ -165,7 +165,8 @@ class DependencyManager {
         ControlElements sampleModifyControls
         try {
-            sampleModifyControls = new ControlElements(this.portletController, this.viewModel)
+            def userEmail = PortalUtils.isLiferayPortlet() ? PortalUtils.getUser().getEmailAddress() : "Not logged in."
```
I like that!

KochTobi commented on src/main/groovy/life/qbic/portal/sampletracking/datasources/SampleTracker.groovy:
```diff
@@ -1,6 +1,7 @@
 package life.qbic.portal.sampletracking.datasources
 
 import groovy.util.logging.Log4j2
+import io.micronaut.core.type.Argument
```
Where does that `Argument` come from? What does it do? Let's have an argument... (sorry couldn't resist)

sven1103 replied:
That's what it does. It enables that a complex expected response object can be converted properly:
```
try {
                response = client.withCloseable { it.toBlocking().exchange(request,
                                                    Argument.of(List.class, Location.class))}
   }
```

KochTobi commented on src/main/groovy/life/qbic/portal/sampletracking/datasources/SampleTracker.groovy:
```diff
@@ -78,23 +79,33 @@ class SampleTracker {
-            if (!response?.body() instanceof List) {
-                throw new SampleTrackingQueryException("Did not receive a valid List response.")
+
+            if (response.getBody().empty()) {
```
Quite the improvement. Way more readable. We should do it like this in every comparable method.

KochTobi commented on src/main/groovy/life/qbic/portal/sampletracking/web/ViewModel.groovy:
```diff
@@ -11,6 +11,10 @@ class ViewModel {
     final ObservableList successNotifications
     final ObservableList failureNotifications
 
+    ViewModel() {
```
Yeah why not.

KochTobi commented on src/main/groovy/life/qbic/portal/sampletracking/web/presenters/ControlElementsPresenter.groovy:
```diff
@@ -22,11 +22,11 @@ class ControlElementsPresenter implements QueryAvailableLocationsOutput {
     @Override
     def invokeOnError(String msg) {
-        return null
+        this.viewModel.failureNotifications << msg
```
I like the `.add` method better since it is more readable and common in Java programs.
```suggestion
        this.viewModel.failureNotifications.add(msg)
```

KochTobi commented on src/main/groovy/life/qbic/portal/sampletracking/web/views/ControlElements.groovy:
```diff
@@ -20,29 +21,34 @@ class ControlElements extends VerticalLayout {
+        this.userEmail = userEmail
         initLayout()
         registerListeners()
+        portletController.queryAllLocationsForPerson(userEmail)
```
We could also move this out of the constructor and to a FocusListener. That would enable the creation of an empty View. But since the service should always be present this could also be ok.
Calling the method in the FocusListener and presenting the user with the message "Currently not possible. Try again later" decouples the availability of the portlet itself from the service. That is a good thing I guess.

KochTobi reviewed (changes requested):
Some minor comments I would like to have a discussion about before merge.

KochTobi committed:
Merge branch 'code-refractor' into feature/usecase-controlelements

KochTobi commented on src/main/groovy/life/qbic/portal/sampletracking/web/views/ControlElements.groovy:
```diff
@@ -20,29 +21,34 @@ class ControlElements extends VerticalLayout {
-    ControlElements(PortletController portletController, ViewModel viewModel) {
+    private ControlElements() {
+        //prevent default constructor initialization
+        throw new AssertionError("${ControlElements.getSimpleName()} cannot be initialized by the its default constructor.")
```
```suggestion
        throw new AssertionError("${ControlElements.getSimpleName()} cannot be initialized by its default constructor.")
```

KochTobi commented:
Maybe rename the branch or create a new branch after this PR is merged. I find it useful when small features get their own branch and massive PRs (and therefore probably merge conflicts) are avoided.
For example I would have called this branch `feature/show-possible-locations`
Another branch for the ControlElements could be `feature/control-elements-functionality`

sven1103 committed:
Use method name over list operator

Co-Authored-By: KochTobi <KochTobi@users.noreply.github.com>

sven1103 committed:
Corrects assertion error message

Co-Authored-By: KochTobi <KochTobi@users.noreply.github.com>

sven1103 committed:
Prevents null from beeing displayed in the select menu

sven1103 committed:
Merge branch 'code-refractor' into feature/usecase-controlelements

sven1103 committed:
Passes selected location and date to use case

sven1103 committed:
Merge branch 'feature/usecase-controlelements' of github.com:qbicsoftware/sample-tracking-update-portlet into feature/usecase-controlelements

sven1103 committed:
Sets the location and status for the update

When the update sample list is pressed, the selected samples
get updated to the selected location and status

sven1103 committed:
Calls the presenter to execute the update finish routine

sven1103 committed:
Remove successfully updated samples from list

sven1103 committed:
Merge branch 'code-refractor' into feature/usecase-controlelements

KochTobi reviewed (changes requested):
Input from @wow-such-code required on specification.
Otherwise looks good.

sven1103 committed:
Use PUT over POST to update status and/or location

sven1103 committed:
Merge branch 'feature/usecase-controlelements' of github.com:qbicsoftware/sample-tracking-update-portlet into feature/usecase-controlelements

sven1103 commented on src/main/groovy/life/qbic/portal/sampletracking/web/controllers/SampleTrackingPortletController.groovy:
```diff
@@ -48,9 +51,13 @@ class SampleTrackingPortletController implements PortletController {
+
         for (sampleId in sampleIds) {
-            this.sampleUpdateInput.setCurrentSampleLocation(sampleId, desiredLocation)
+            this.sampleUpdateInput.setCurrentSampleLocation(sampleId, location)
```
addressed and solved in b36dc69

sven1103 committed:
Merge branch 'code-refractor' into feature/usecase-controlelements

KochTobi committed:
Move available locations query out of constructor to DependencyManager

Previously the constructor in the ControlElements view called the controller to update the available locations list
This is not moved to the DependencyManager. The portlet fails there if no controller exists and no longer fails
within the constructor of a "studid" view class.

KochTobi commented:
Works like a charm now
